adam/learner/alignments.py

```diff
@@ -0,0 +1,310 @@
+        # replacing any spans of text which are aligned to objects with variables.
+        # We iterate backwards so we the indices don't get invalidated
+        # as we replace parts of the token sequence with template variables.
+        object_nodes_sorted_by_reversed_aligned_token_position = tuple(
+            reversed(
+                sorted(
+                    node_to_language_span_restricted.keys(),
+                    key=lambda match_node: node_to_language_span_restricted[match_node],
+                )
+            )
+        )
+
+        for matched_object_node in object_nodes_sorted_by_reversed_aligned_token_position:
+            aligned_token_span = node_to_language_span_restricted[matched_object_node]
+            # If an object is aligned to multiple tokens,
+            # we just delete all but the first.
+            if len(aligned_token_span) > 1:
+                for non_initial_aligned_index in range(
+                    # -1 because the end index is exclusive
+                    aligned_token_span.end - 1,
+                    aligned_token_span.start,
+                    -1,
+                ):
+                    del template_elements[non_initial_aligned_index]
+            # Regardless, the first token is replaced by a variable.
+            template_elements[
+                aligned_token_span.start
+            ] = object_node_to_template_variable[matched_object_node]
+
+        return SurfaceTemplate(
+            template_elements, determiner_prefix_slots=determiner_prefix_slots
+        )
+
+
+@attrs(frozen=True)
+class PerceptionSemanticAlignment:
+    perception_graph: PerceptionGraph = attrib(validator=instance_of(PerceptionGraph))
+    semantic_nodes: ImmutableSet[SemanticNode] = attrib(
+        converter=_to_immutableset, validator=deep_iterable(instance_of(SemanticNode))
+    )
+
+    @staticmethod
+    def create_unaligned(
+        perception_graph: PerceptionGraph
+    ) -> "PerceptionSemanticAlignment":
+        return PerceptionSemanticAlignment(perception_graph, [])
+
+    def copy_with_updated_graph_and_added_nodes(
+        self, *, new_graph: PerceptionGraph, new_nodes: Iterable[SemanticNode]
+    ) -> "PerceptionSemanticAlignment":
+        if new_graph is self.perception_graph and not new_nodes:
+            return self
+        else:
+            return PerceptionSemanticAlignment(
+                perception_graph=new_graph,
+                semantic_nodes=chain(self.semantic_nodes, new_nodes),
+            )
+
+    def __attrs_post_init__(self) -> None:
+        for node in self.perception_graph._graph:
+            if isinstance(node, SemanticNode):
+                if not node in self.semantic_nodes:
+                    raise RuntimeError(
+                        "All semantic nodes appearing in the perception graph must "
+                        "also be in semantic_nodes"
+                    )
+
+
+@attrs(frozen=True, kw_only=True)
+class LanguagePerceptionSemanticAlignment:
```

Comment:
📖

Reply:
@gabbard Done.

Also, I noticed that the docstring on `LanguageConceptAlignment` seems to be [out of date][old_docstring]. It's written as if the class represented a direct alignment between language and perception whereas now it's between language and concepts. Should I update that as well?

[old_docstring]: https://github.com/isi-vista/adam/blob/91d5bb7bb9c2e0a0e8679e52618f2468be76d391/adam/learner/alignments.py#L35

Reply:
Please do

Reply:
@gabbard [Done](https://github.com/isi-vista/adam/blob/787375cb1a21b3a0c1a3d92ed456ee22802dbc9a/adam/learner/alignments.py#L35).